Suboptimal error in case of duplicate `,` in struct constructor

```rust
struct Foo {
    a: i32,
    b: i32,
}

fn main() {
    let x = Foo {
        a: 42,,
        b: 3,
    };
}
```

results  in 

```
error: expected identifier, found `,`
 --> src/main.rs:9:15
  |
8 |     let x = Foo {
  |             --- while parsing this struct
9 |         a: 42,,
  |               ^ expected identifier

error[E0063]: missing field `b` in initializer of `Foo`
 --> src/main.rs:8:13
  |
8 |     let x = Foo {
  |             ^^^ missing `b`
```

we may want to add a parser recovery that just reports the error, eats the comma and continues parsing as normally

--- a/src/libsyntax/parse/parser.rs
+++ b/src/libsyntax/parse/parser.rs
@@ -776,7 +776,7 @@ impl<'a> Parser<'a> {
             err.span_label(self.span, format!("expected identifier, found {}", token_descr));
         } else {
             err.span_label(self.span, "expected identifier");
-            if self.token == token::Comma && self.look_ahead(1, |t| *t.is_ident()) {
+            if self.token == token::Comma && self.look_ahead(1, |t| t.is_ident()) {
                 err.span_suggestion(self.span, "remove this comma", "".into());
             }
         }

--- a/src/libsyntax/parse/parser.rs
+++ b/src/libsyntax/parse/parser.rs
@@ -776,6 +776,9 @@ impl<'a> Parser<'a> {
             err.span_label(self.span, format!("expected identifier, found {}", token_descr));
         } else {
             err.span_label(self.span, "expected identifier");
+            if self.token == token::Comma {
+                err.span_suggestion(self.span, "try removing a comma", ",".into());
+            }
         }
         err
     }
@@ -2446,8 +2449,11 @@ impl<'a> Parser<'a> {
                 Err(mut e) => {
                     e.span_label(struct_sp, "while parsing this struct");
                     e.emit();
-                    self.recover_stmt();
-                    break;
+
+                    if self.token != token::Comma {
+                        self.recover_stmt();
+                        break;
+                    }
                 }
             }
 

--- a/src/libsyntax/parse/parser.rs
+++ b/src/libsyntax/parse/parser.rs
@@ -2450,6 +2450,9 @@ impl<'a> Parser<'a> {
                     e.span_label(struct_sp, "while parsing this struct");
                     e.emit();
 
+                    // If the next token is a comma, then try to parse
+                    // what comes next as additional fields, rather than
+                    // bailing out until next `}`.
                     if self.token != token::Comma {
                         self.recover_stmt();
                         break;

--- a/src/libsyntax/parse/parser.rs
+++ b/src/libsyntax/parse/parser.rs
@@ -777,7 +777,7 @@ impl<'a> Parser<'a> {
         } else {
             err.span_label(self.span, "expected identifier");
             if self.token == token::Comma {
-                err.span_suggestion(self.span, "try removing a comma", ",".into());
+                err.span_suggestion(self.span, "remove this comma", ",".into());
             }
         }
         err

--- a/src/libsyntax/parse/parser.rs
+++ b/src/libsyntax/parse/parser.rs
@@ -776,8 +776,8 @@ impl<'a> Parser<'a> {
             err.span_label(self.span, format!("expected identifier, found {}", token_descr));
         } else {
             err.span_label(self.span, "expected identifier");
-            if self.token == token::Comma {
-                err.span_suggestion(self.span, "remove this comma", ",".into());
+            if self.token == token::Comma && self.look_ahead(1, |t| *t.is_ident()) {
+                err.span_suggestion(self.span, "remove this comma", "".into());
             }
         }
         err

--- a/src/test/ui/struct-duplicate-comma.rs
+++ b/src/test/ui/struct-duplicate-comma.rs
@@ -1,0 +1,25 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// compile-flags: -Z parse-only
+
+struct Foo {
+    a: u8,
+    b: u8
+}
+
+fn main() {
+    let bar = Foo {
+        a: 0,,
+          //~^ ERROR expected identifier
+        b: 42
+    };
+}
+

--- a/src/test/ui/struct-duplicate-comma.rs
+++ b/src/test/ui/struct-duplicate-comma.rs
@@ -10,6 +10,8 @@
 
 // compile-flags: -Z parse-only
 
+// Issue #50974
+
 struct Foo {
     a: u8,
     b: u8

--- a/src/test/ui/struct-duplicate-comma.stderr
+++ b/src/test/ui/struct-duplicate-comma.stderr
@@ -1,0 +1,13 @@
+error: expected identifier, found `,`
+  --> $DIR/struct-duplicate-comma.rs:20:14
+   |
+LL |     let bar = Foo {
+   |               --- while parsing this struct
+LL |         a: 0,,
+   |              ^
+   |              |
+   |              expected identifier
+   |              help: remove this comma: `,`
+
+error: aborting due to previous error
+

--- a/src/test/ui/struct-duplicate-comma.stderr
+++ b/src/test/ui/struct-duplicate-comma.stderr
@@ -1,5 +1,5 @@
 error: expected identifier, found `,`
-  --> $DIR/struct-duplicate-comma.rs:20:14
+  --> $DIR/struct-duplicate-comma.rs:22:14
    |
 LL |     let bar = Foo {
    |               --- while parsing this struct

--- a/src/test/ui/struct-duplicate-comma.stderr
+++ b/src/test/ui/struct-duplicate-comma.stderr
@@ -7,7 +7,7 @@ LL |         a: 0,,
    |              ^
    |              |
    |              expected identifier
-   |              help: remove this comma: `,`
+   |              help: remove this comma
 
 error: aborting due to previous error
 

